rpki-validator/src/main/java/net/ripe/rpki/validator3/util/HappyEyeballsResolver.java

```diff
@@ -0,0 +1,256 @@
+                log.debug("Resolved v6 addresses: {}",
+                        resolvedAddressesV6.stream().map(InetAddress::getHostAddress).collect(Collectors.joining(", ")));
+                log.debug("Resolved v4 addresses: {}",
+                        resolvedAddressesV4.stream().map(InetAddress::getHostAddress).collect(Collectors.joining(", ")));
+            }
+
+            final ConcurrentLinkedQueue<SocketChannel> sockets = new ConcurrentLinkedQueue<>();
+            final Thread connectionsThread =
+                    new Thread(connectAttemptsRunnable(resolvedAddressesV6, resolvedAddressesV4, sockets, port));
+            connectionsThread.start();
+
+            bestAddress = (InetSocketAddress) awaitSuccessfulConnection(sockets);
+            if (bestAddress == null) {
+                promise.failed(new UnknownHostException(host));
+            } else {
+                promise.succeeded(Collections.singletonList(bestAddress));
+            }
+
+            connectionsThread.interrupt();
+            v6thread.interrupt();
+            v4thread.interrupt();
+            for (SocketChannel channel : sockets) {
+                try {
+                    channel.close();
+                } catch (Exception e) {
+                    // ignore
+                }
+            }
+        } catch (Exception e) {
+            log.warn("Error during looking for happy eyeballs: ", e);
+            promise.failed(e);
+        }
+    }
+
+    private static SocketAddress awaitSuccessfulConnection(ConcurrentLinkedQueue<SocketChannel> sockets) throws IOException {
+        boolean keepGoing = true;
+        while (keepGoing) {
+            final Iterator<SocketChannel> iterator = sockets.iterator();
+            int count = 0;
+            while (iterator.hasNext()) {
+                count++;
+                final SocketChannel socketChannel = iterator.next();
+                if (socketChannel == SOCKETS_TERMINATOR) {
+                    if (count == 1) {
+                        keepGoing = false;
+                    }
+                    break;
+                }
+                try {
+                    if (socketChannel.finishConnect()) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("Successfully connected to: {}", socketChannel.getRemoteAddress());
+                        }
+                        return socketChannel.getRemoteAddress();
+                    }
+                } catch (IOException e) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Error connecting to " + socketChannel.getRemoteAddress(), e);
+                    }
+                    iterator.remove();
+                }
+            }
+        }
+        return null;
+    }
+
+    private static void awaitDnsResponses(ConcurrentLinkedQueue<InetAddress> resolvedAddressesV6,
+                                          ConcurrentLinkedQueue<InetAddress> resolvedAddressesV4) {
```

Comment:
Just trying to understand how this work. Before entering this await block we already spawn v4 and v6 lookup threads. Essentially this block tries to wait until we have any v6, with some deadline. 

What I am not sure I understand is that the flag to stop going. It is now within a block of if v4 is not empty. 

So it kinda assume that the lookup thread for v4 always returns a value (otherwise once entering the await, it won't stop). Do we have guarantee this loop won't stay when v4 lookup returns nothing?

Reply:
The logic according to the RFC should be:
- if the v6 address comes first, use it first
- if the v4 address comes first, wait 50ms more
- if v6 address comes within those 50 ms, use it first
- otherwise use v4 address first

Reply:
And, both lookups will return `ADDRESSES_TERMINATOR` at the end of their queues, so both `!isEmpty` conditions should become true at some point in time.